R/method-filterFeatures.R

```diff
@@ -0,0 +1,597 @@
+#'
+#' @return For `BlankFlag`: a `BlankFlag` class. `filterFeatures` returns
+#' the input object with an added column in the features metadata called
+#' `possible_contaminants` with a logical value for each feature. This is added
+#' to `featureDefinitions` for `XcmsExperiment` objects and `rowData` for
+#' `SummarizedExperiment` objects.
+#'
+#' @author Philippine Louail
+#'
+#' @importFrom MetaboCoreUtils rowBlank
+#'
+NULL
+
+#' @noRd
+setClass("BlankFlag",
+         slots = c(threshold = "numeric", blankIndex = "integer",
+                   qcIndex = "integer", na.rm = "logical"),
+         contains = "Param",
+         prototype = prototype(
+             threshold = numeric(),
+             blankIndex = integer(),
+             qcIndex = integer(),
+             na.rm = logical()),
+         validity = function(object) {
+             msg <- NULL
+             if (length(object@threshold) != 1)
+                 msg <- c("'threshold' has to be a numeric of length 1")
+             msg
+         })
+
+#' @rdname BlankFlag
+#'
+#' @export
+BlankFlag <- function(threshold = 2, blankIndex = integer(),
+                         qcIndex = integer(), na.rm = TRUE) {
+    if (is.logical(blankIndex))
+        blankIndex <- which(blankIndex)
+    if(is.numeric(blankIndex))
+        blankIndex <- as.integer(blankIndex)
+    if (is.logical(qcIndex))
+        qcIndex <- which(qcIndex)
+    if(is.numeric(qcIndex))
+        qcIndex <- as.integer(qcIndex)
+    new("BlankFlag", threshold = threshold, blankIndex = blankIndex,
+        qcIndex = qcIndex, na.rm = na.rm)
+}
+
+#' @rdname BlankFlag
+setMethod("filterFeatures",
+          signature(object = "XcmsResult",
+                    filter = "BlankFlag"),
+          function(object, filter){
+              if (length(filter@blankIndex) < 1 |
+                  length(filter@blankIndex) > length(object))
+                  stop("'blankIndex' must be within object length range")
+              if (length(filter@qcIndex) < 1 |
+                  length(filter@qcIndex) > length(object))
+                  stop("'qcIndex' must be within object length range")
+              x <- featureValues(object)[, filter@qcIndex]
+              y <- featureValues(object)[, filter@blankIndex]
+              vals <- rowBlank(x = x, y = y,
+                               na.rm = filter@na.rm,
+                               threshold = filter@threshold)
+              message(paste(length(featureValues(object)) - sum(vals),
+                            "features were flagged"))
+              featureDefinitions(object)$possible_contaminants <- vals
+              ph <- XProcessHistory(param = filter,
+                                    date. = date(),
+                                    type. = .PROCSTEP.FEATURE.FILTERING,
+                                    fileIndex = seq_along(object))
+              object <- addProcessHistory(object, ph)
+              validObject(object)
+              object
```

Comment:
here do I need to call the object at the end, or validObject() will return the object anyway ? I think it won't work but just checking